internal/controllers/topology/cluster/desired_state.go

```diff
@@ -427,10 +447,45 @@ func computeCluster(_ context.Context, s *scope.Scope, infrastructureCluster, co
 
 	// Set the references to the infrastructureCluster and controlPlane objects.
 	// NOTE: Once set for the first time, the references are not expected to change.
-	cluster.Spec.InfrastructureRef = contract.ObjToRef(infrastructureCluster)
-	cluster.Spec.ControlPlaneRef = contract.ObjToRef(controlPlane)
+	var err error
+	cluster.Spec.InfrastructureRef, err = calculateRefDesiredAPIVersion(cluster.Spec.InfrastructureRef, infrastructureCluster)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate infrastructureRef")
+	}
+	cluster.Spec.ControlPlaneRef, err = calculateRefDesiredAPIVersion(cluster.Spec.ControlPlaneRef, controlPlane)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate controlPlaneRef")
+	}
+
+	return cluster, nil
+}
+
+// calculateRefDesiredAPIVersion returns the desired ref calculated from desiredReferencedObject
+// so it doesn't override the version in apiVersion stored in the currentRef, if any.
+// This is required because the apiVersion in the desired ref is aligned to the apiVersion used
+// in ClusterClass when reading the current state. If the currentRef is nil or group or kind
+// doesn't match, no changes are applied to desired ref.
+func calculateRefDesiredAPIVersion(currentRef *corev1.ObjectReference, desiredReferencedObject *unstructured.Unstructured) (*corev1.ObjectReference, error) {
```

Comment:
q: should we apply this func for MD templates as well?

Reply:
Nice catch. Currently the MD controller does not bump the refs so it's not an issue (because of a bug). But let's make it more robust against changes of the MD controller even if we're not sure if we will ever introduce the bump there.